### ISSUE TITLE
Early Stopping + result dictionary + no validation not working.

### ISSUE THREAD

**Issue by Lucas-Steinmann** [PyTorchLightning/pytorch-lightning#3199]
## 🐛 Bug

The case where the user does not use validation __and__ returns a dictionary (instead of a TrainResult) during training does not work in combination with early stopping.
The test case which should check this is here:

https://github.com/PyTorchLightning/pytorch-lightning/blob/bd35c869ee0b856fc30229ef36957b058fe97461/tests/callbacks/test_early_stopping.py#L136-L159

The check in the last line is wrong. It should actually compare:
```python3
    assert trainer.current_epoch < trainer.max_epochs - 1
```

### To Reproduce

1. Fix the test case
2. Run tests.

#### Code sample

I guess using the test case is simpler and easier.

### Expected behavior

That is an interesting question indeed. Possibilities are:

1. Test case should pass with correct comparison
2. The [docs](https://github.com/PyTorchLightning/pytorch-lightning/blob/bd35c869ee0b856fc30229ef36957b058fe97461/docs/source/results.rst#checkpointearly_stop) and  @williamFalcon in https://github.com/PyTorchLightning/pytorch-lightning/pull/3193#discussion_r477245884 suggest that only 'loss' should work.

So before fixing this issue, it should be settled what the expected behavior is.
If you tell me, I'm happy to help.
I could also include it in the pull request where I already tried to bring the docs in line with the test cases.

### Environment

* CUDA:
	- GPU:
		- GeForce GTX 1080 Ti
	- available:         True
	- version:           10.1
* Packages:
	- numpy:             1.19.1
	- pyTorch_debug:     False
	- pyTorch_version:   1.6.0
	- pytorch-lightning: 0.9.1dev
	- tensorboard:       2.2.0
	- tqdm:              4.48.2
* System:
	- OS:                Linux
	- architecture:
		- 64bit
		- ELF
	- processor:         x86_64
	- python:            3.8.3
	- version:           #113~16.04.1-Ubuntu SMP Fri Jul 10 04:37:08 UTC 2020

**Comment by ananyahjha93** [comment]
@Lucas-Steinmann mind sending a PR for this?

**Comment by Lucas-Steinmann** [comment]
I'd love to, but what is the intended behavior? 
Possibility 1, 2, or doesn't it matter? 
Maybe we should wait on @williamFalcon answer, since he already expressed his opinion.

**Comment by Lucas-Steinmann** [comment]
There is a third possibility, which is suggested by the code which would have to be changed:
https://github.com/PyTorchLightning/pytorch-lightning/blob/4d98419bb84e06c06aaee0a86530e322e24f6d11/pytorch_lightning/callbacks/early_stopping.py#L169

This comment says, that the current behavior is correct. (assuming "structured result" means Result object)
So either code, docs or test is correct and the other two have to be adjusted. :)

**Comment by williamFalcon** [comment]
if you're not using the trainresult then the key 'loss' needs to be present

All of these are equivalent

```python

return loss

return {'loss': loss}

return TrainResult(minimize=loss)

return TrainResult(loss)
```

**Comment by Lucas-Steinmann** [comment]
But it doesn't work ``return {'loss': loss}`` with early stopping.

Also the test, which suggests that any key should work still contradicts you?

Furthermore the test is still wrong because the last line does not actually check whether early stopping was triggered (**off-by-one error**)

**Comment by ananyahjha93** [comment]
@Lucas-Steinmann reopening this then!

**Comment by Lucas-Steinmann** [comment]
Just to be clear. The next step is for you (the project owners) to decide what the correct behaviour is:

1. The docs (only dicts with 'loss' key should work with early stopping): Then we have to fix the code and the tests.
2. The code (early stopping should only work with TrainResult): Then we have to remove the test and fix the docs
3. The test (early stopping should work with dicts any key): Then we have to fix the code and the docs.

I would love to help but I think it is not for me to decide what is correct.
If you pick one, I'll implement it.

**Comment by williamFalcon** [comment]
The correct behavior is neither of the above...

1. if using dicts, it should work with whatever you put in the monitor argument of the EarlyStopping or ModelCheckpoint.

```python
EarlyStopping(monitor='jiraffe')

# must have this key
return {'jiraffe': x}
```

2. If using results it should work with whatever is in the early_stop_on argument:

```python
acc = ...
loss = ...
f1 = ...

# pick whatever you want to early stop on
key_i_care_about = f1 or loss or acc
result = pl.TrainResult(early_stop_on=key_i_care_about)
```

If you ALSO pass the key in the EvalResult then the EvalResult takes precedence...

**Comment by Lucas-Steinmann** [comment]
Ok. 1. currently does not worked (atleast a few days ago, when I wrote the issue, see my minimal example here: https://github.com/PyTorchLightning/pytorch-lightning/pull/3193#discussion_r478517184)

There is a test, which exactly checks 1.  but has an off-by-one error and would actually fail if the error is fixed (see my first comment on this issue).

If I understand the docs correctly, they currently say it should only work with 'loss' key. Maybe it was meant to say the default behaviour is that it only works with 'loss' key: https://github.com/PyTorchLightning/pytorch-lightning/blob/3910ad033074367f6abfe0001562db725a75cb73/docs/source/results.rst#L109

**Comment by Lucas-Steinmann** [comment]
To prevent misunderstanding:
```
EarlyStopping(monitor='jiraffe')

# must have this key
return {'jiraffe': x}
```
is without validation step right, i.e. the ``return {'jiraffe': x}`` is in ``training_step()``?

I've implemented a pull request. #3347.
It first corrects the test, which then fails. Then I fixed the issue.
Afterwards another bug was unveiled, which I fixed in the subsequent commits.
Then I also updated the docs.
